### PR TITLE
Python bindings for ExtractPitch and Transient Detection Helper Funcs

### PR DESCRIPTION
**Pull request by mabruzzo**
- Wrote python bindings for `ExtractPitch` and helper functions used to compute the detection function used by the pairwise transient detection algorithm (`centralFreqMapper`, `rollSigma`, `pSMContribution`, `sosGammatone`)
- Tweaked the function signature of and wrote documentation for `centralFreqMapper`, `rollSigma`, and `pSMContribution`
- Removed the unused max_capacity field from `intList` and `distinctList`
- Cleaned up lists.c - mainly moved some commented-out throwaway tests for `orderedList` and `distinctList` to a formal unit test

### CODE REVIEW

File: src/filterBank.c [hunk 1]
```diff
@@ -79,64 +81,14 @@ void filterBankDestroy(struct filterBank* fB){
...
-	 * The ith entry of fcArray will have ERBS given by
-	 * ERBS(fcArray[i]) = minERBS + i * (maxERBS-minERBS)/(numChannels - 1)
-	 *
-	 * We can calculate the ith entry of fcArray by taking the inverse of 
-	 * the ERBS of the ith entry
-	 * fcArray[i] =(10.^((minERBS + i * (maxERBS-minERBS)/(numChannels - 1))
-	 *                   /21.4) - 1)/0.00437
-	 */
-	float *fcArray, minERBS,maxERBS;
+int centralFreqMapper(int numChannels, float minFreq, float maxFreq,
```
Comment (richardsp256) [inline src/filterBank.c]: Why does this return an int? it should just return void, the return value isn't used for anything.

File: src/simpleDetFunc.c [hunk 2]
```diff
@@ -130,15 +130,15 @@ static inline void remove_var(double val, int *nobs, double *mean_x,
...
+ * python pandas. This algorithm can still be further optimized so that we add
+ * and remove values from the windows in chunks (rather than 1 at a time). 
+ * Additionally, if we store wInd on the stack, we might get better 
+ * optimizations.
  */
 
-void rollSigma(int startIndex, int interval, float scaleFactor,
-	       int sigWindowSize, int dataLength, int numWindows,
-	       float *buffer, float *sigmas)
+int rollSigma(int startIndex, int interval, float scaleFactor,
```
Comment (richardsp256) [inline src/simpleDetFunc.c]: Same as centralFreqMapper, why was this changed to return an int instead of void?

File: src/simpleDetFunc.c [hunk 3]
```diff
@@ -263,9 +264,9 @@ static inline float calcPSMEntryContrib(float* x, int window_size, float sigma)
 	return out;
 }
 
-void pSMContribution(int correntropyWinSize, int interval,
-		     int numWindows, float *buffer,
-		     float *sigmas,float **pSMatrix)
+int pSMContribution(int correntropyWinSize, int interval,
```
Comment (richardsp256) [inline src/simpleDetFunc.c]: Ditto on return value

File: src/simpleDetFunc.c [hunk 4]
```diff
@@ -376,21 +378,21 @@ int simpleDetFunctionCalculation(int correntropyWinSize, int interval,
 	 * function that computes correntropy. Add correntropyWinSize zeros 
 	 * will be sufficient.
 	 */
-	bufferLength = dataLength + correntropyWinSize;
+	bufferLength = (numWindows-1)*interval + 2*correntropyWinSize + 2;
```
Comment (richardsp256) [inline src/simpleDetFunc.c]: just wondering why this change was made and how you determined this new bufferlength. updating the comment above this line is probably a good idea.

File: src/simpleDetFunc.c [hunk 5]
```diff
@@ -376,21 +378,21 @@ int simpleDetFunctionCalculation(int correntropyWinSize, int interval,
...
 
-	
 	pooledSummaryMatrix = malloc(sizeof(float)*numWindows);
 	for (i=0;i<numWindows;i++){
 		pooledSummaryMatrix[i] = 0;
 	}
 	sigmas = malloc(sizeof(float)*numWindows);
 
-	centralFreq = centralFreqMapper(numChannels, minFreq, maxFreq);
+	centralFreq = malloc(sizeof(float)*numChannels);
```
Comment (richardsp256) [inline src/simpleDetFunc.c]: Whenever there is a malloc call you should always check if NULL is returned, and return an error accordingly. You do it when calling this same function in filterbank.c in this commit.

File: src/lists.h [hunk 6]
```diff
@@ -27,7 +27,6 @@ typedef struct {
 	struct distinctCandidate *array;
 	int length;
 	int capacity;
-	int max_capacity;
```
Comment (richardsp256) [inline src/lists.h]: Whoops, I missed this when removing max_capacity :(  thanks for catching this.

File: tests/CMakeLists.txt [hunk 7]
```diff
@@ -39,8 +39,15 @@ set(DETFUNCTION_TEST_SOURCES
...
 )
 
+set(LISTS_TEST_SOURCES
+  check_lists.c
+)
+
 add_executable(check_gammatone ${GAMMATONE_TEST_SOURCES} ${ARRAY_TEST_SOURCES})
 add_executable(check_detFunction ${DETFUNCTION_TEST_SOURCES}
   ${ARRAY_TEST_SOURCES})
+add_executable(check_lists ${LISTS_TEST_SOURCES})
```
Comment (richardsp256) [inline tests/CMakeLists.txt]: If I am understanding this correctly, are you making a different executable for each test file? It doesn't need to be changed for this commit, but if that is the case down the line I think it would be better to have a single test executable that runs all tests. Once we start adding lots of tests it will become a real pain otherwise.

File: src/filterBank.c (hunk 1 above)
Comment (mabruzzo) [reply src/filterBank.c]: This was a dumb thing I did on my part as something of a stopgap solution. While writing the python bindings, the ctypes documentation didn't mention how to specify that a function returns a void value... I since found a stack overflow page explaining how to do that and have reverted the function signatures

File: src/simpleDetFunc.c (hunk 3 above)
Comment (mabruzzo) [reply src/simpleDetFunc.c]: Same as my response for centralFreqMapper - it has been fixed

File: src/simpleDetFunc.c (hunk 4 above)
Comment (mabruzzo) [reply src/simpleDetFunc.c]: I calculated it based on the logic in pSMContribution. I had updated the documentation to state that the buffer argument was assumed to have a length of at least `(numWindows-1)*interval + 2*correntropyWinSize + 2` and confirmed that it was accurate with a test. 
The old length was just a conservative guess at the required length.

In the final commit added to the pull request, I updated the comment.

File: tests/CMakeLists.txt (hunk 7 above)
Comment (mabruzzo) [reply tests/CMakeLists.txt]: I hadn't given this a ton of thought and was just following the simple tutorial that they provide.

The basic idea in the check framework is that you organize test cases into suites and you might have a few suites in a file. I think the idea is for test suites to consolidate test cases that have common setup and teardown steps. I generally agree that having a ton of files wouldn't be great. There are 2 tradeoffs to having a single file executable:

1. We need to deal with header files (this isn't a problem)
2. It becomes harder to selectively run a set of tests. It is doable to [selectively run a test suite](https://libcheck.github.io/check/doc/check_html/check_4.html#Selecting-Tests-by-Suite-or-Test-Case). This probably isn't much of an issue for now (since it doesn't take much time to run everything). However in the future, we might want to be careful how we group together test suites (e.g. we have more than one test suite for gammatone filters that each hold test cases that test different aspects of the implementation). 

To summarize: I agree that it would be useful to have a single executable, but we should be somewhat mindful that if we have a lot of tests, we will need to be able to intelligently specify subsets of tests to run.
